### PR TITLE
[GHF] Add submodule updates check

### LINKED ISSUE #74326: Lint rule that would fail for PRs containing subdmodule updates
### 🚀 The feature, motivation and pitch

When submodules gets updated after checkout, devs often forget to run `git submodule update` command and propose PR that effectively reverts the update.

Prevent those accidental updates by requiring one to add "submodule-update" keyword to PR description before change can be merged.

### Alternatives

_No response_

### Additional context

_No response_

cc @seemethere @malfet @pytorch/pytorch-dev-infra

### PR DESCRIPTION
**Pull request by malfet**
Originally planned to integrate it somehow into the `lintrunner`, but this poses too many challenges, one of them is that it deliberately ignores submodule updates.

On the other hand, almost all the information, other than list of the submodules is already present in the GitHubPR info.

Incorporate small BE change into `test_trymerge.py`, that moves `@mock.patch` from individual test to the class definition.

Fixes https://github.com/pytorch/pytorch/issues/74326 and https://github.com/pytorch/test-infra/issues/1521

### CODE REVIEW

File: .github/scripts/trymerge.py [hunk 1]
```diff
@@ -725,6 +742,29 @@ def get_changed_files(self) -> List[str]:
...
+                            owner=self.org)
+            info = rc["data"]["repository"]["submodules"]
+            self.submodules = [s["path"] for s in info["nodes"]]
+        return self.submodules
+
+    def get_changed_submodules(self) -> List[str]:
+        submodules = self.get_submodules()
+        return [f for f in self.get_changed_files() if f in submodules]
+
+    def has_valid_submodule_updates(self) -> bool:
```
Comment (ZainRizvi) [inline .github/scripts/trymerge.py:758]: nit: feels weird for this to return true when there are zero submodule updates.

Might be better to rename it to `has_invalid_submodule_updates`

**Review by ZainRizvi — APPROVED** [review]

File: .github/scripts/trymerge.py (hunk 1 above)
Comment (malfet) [reply .github/scripts/trymerge.py:758]: Sure

**Review by huydhn — APPROVED** [review]

File: .github/scripts/trymerge.py [hunk 2]
```diff
@@ -725,6 +742,29 @@ def get_changed_files(self) -> List[str]:
...
+            self.submodules = [s["path"] for s in info["nodes"]]
+        return self.submodules
+
+    def get_changed_submodules(self) -> List[str]:
+        submodules = self.get_submodules()
+        return [f for f in self.get_changed_files() if f in submodules]
+
+    def has_invalid_submodule_updates(self) -> bool:
+        """ Submodule updates in PR are invalid if submodule keyword
+        is not mentioned in neither the title nor body/description
```
Comment (ZainRizvi) [inline .github/scripts/trymerge.py:760]: nit

```suggestion
        is neither mentioned in the title nor body/description
```

**Review by ZainRizvi — APPROVED** [review]

**Review by alexsio27444 — CHANGES_REQUESTED** [review]
Change

**Pull request closed without merging**